sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala

```diff
@@ -46,35 +46,42 @@ object AgnosticEncoders {
     override val clsTag: ClassTag[Option[E]] = ClassTag(classOf[Option[E]])
   }
 
-  case class ArrayEncoder[E](element: AgnosticEncoder[E])
+  case class ArrayEncoder[E](element: AgnosticEncoder[E], containsNull: Boolean)
     extends AgnosticEncoder[Array[E]] {
     override def isPrimitive: Boolean = false
-    override def dataType: DataType = ArrayType(element.dataType, element.nullable)
+    override def dataType: DataType = ArrayType(element.dataType, containsNull)
     override val clsTag: ClassTag[Array[E]] = element.clsTag.wrap
   }
 
-  case class IterableEncoder[C <: Iterable[E], E](
+  case class IterableEncoder[C, E](
       override val clsTag: ClassTag[C],
-      element: AgnosticEncoder[E])
+      element: AgnosticEncoder[E],
+      containsNull: Boolean,
+      override val lenientSerialization: Boolean)
```

Comment:
what does lenient mean for a `IterableEncoder`?

Reply:
It means we allow a Seq, a generic Array, or a primitive array as input for serialization

Reply:
Can we leave a code comment to mention it? It's not that obvious compared to `DateEncoder`.

Reply:
yeah will do. TBH I was quite surprised by it.